### PR TITLE
Add support for binary union types - Python 3.10

### PR DESCRIPTION
**Pull request by cdce8p**
## Description
Add basic inference support for runtime binary or unions added in Python 3.10.
Not completely sure this is 100% the right approach for it, but it works. We can always iterate on it later if we want / need to.

At runtime `int | None` is represented as an instance of `types.UnionType`.

Refs https://github.com/PyCQA/pylint/issues/8119

--
After this is merged, I think it would make sense to do a new release so we can include it in pylint `2.16.0`.

### CODE REVIEW

File: astroid/inference.py [hunk 1]
```diff
@@ -814,6 +823,22 @@ def _get_binop_flow(
             _bin_op(left, binary_opnode, op, right, context),
             _bin_op(right, binary_opnode, op, left, reverse_context, reverse=True),
         ]
+
+    if (
+        PY310_PLUS
```
Comment (Pierre-Sassoulas) [inline astroid/inference.py:828]: Isn't it py310 or future annotations used ?

**Review by Pierre-Sassoulas — COMMENTED** [review]
I think this can go in 2.16.

**Comment by Pierre-Sassoulas** [comment]
But shouldn't it be in astroid 2.14 ?

File: astroid/inference.py (hunk 1 above)
Comment (cdce8p) [reply astroid/inference.py:828]: I don't think so. AFAIK inference should model the runtime effects and for Python 3.9 `int | str` would throw a `TypeError`.

We do exclude handle `int | str` in typing contexts with `from __future__ import annotations` separately, in pylint.

**Comment by cdce8p** [comment]
> But shouldn't it be in astroid 2.14 ?

It certainly could. Moved the changelog entry.
We would then only need to decide if we want to release `2.13.4` with the `typing_extensions.TypeVar` fix or just include it with `2.14`.

File: astroid/bases.py [hunk 2]
```diff
@@ -669,3 +670,36 @@ def __repr__(self) -> str:
...
+
+class UnionType(BaseInstance):
+    """Special node representing new style typing unions.
+
+    Proxied class is set once for all in raw_building.
+    """
+
+    _proxied: nodes.ClassDef
+
+    def __init__(self, left, right, parent=None):
```
Comment (DanielNoord) [inline astroid/bases.py:683]: Can we have typing here?

**Review by DanielNoord — CHANGES_REQUESTED** [review]
One minor nit. Rest LGTM!

**Review by DanielNoord — APPROVED** [review]

**Review by Pierre-Sassoulas — CHANGES_REQUESTED** [review]
We should probably cover ``callable``, ``pytype``, ``display_type``, ``bool_value`` and ``__str__`` if only to prevent regression later, right ?

**Comment by cdce8p** [comment]
> We should probably cover `callable`, `pytype`, `display_type`, `bool_value` and `__str__` if only to prevent regression later, right ?

Added some more asserts. Although it might be that we need to go back and change some of them later. I don't know every detail of how the new class will be used by astroid and pylint.

**Comment by cdce8p** [comment]
The isort issue will be fixed by #1989

**Comment by Pierre-Sassoulas** [comment]
:thumbsup:

**Comment by cdce8p** [comment]
> 👍

@Pierre-Sassoulas Think you actually need to approve it since you requested changes earlier. Or I could just bypass the branch protection in this case 😄

**Review by Pierre-Sassoulas — APPROVED** [review]
My bad 😄

**Pull request merged**